### ISSUE TITLE
Please, usage example, I am stupid!

### ISSUE THREAD

**Issue by po0p** [scizzydo/PatternScanner#3]
Hey, nice find ;)
I like the possibilities this library gives!
But please, help me out! I am very stupid, like 70-80 iq below average.
Windows x64 is the target.
Considering this function/byte array:
![x64dbg_1yhuiSdLIY](https://github.com/user-attachments/assets/102b5316-7c6a-4cfb-bb1b-76ddd3649e1b)

I want to read target of the relcall (E8 <4byte displacement>).
Calculated as following: address of E8+1 -> read displacement  as (u)int32_t < not sure if unsigned, probably not
-> target = &(E8)+5+displacement => 00007FFFBAC61099 + 5 + 0x52
![x64dbg_dq8vVSsrIl](https://github.com/user-attachments/assets/0357c606-e34f-4e6f-ad9a-0eb5fe98ec22)

So, my pattern:
```
constexpr auto compiletime_pattern_jmp = "57"
                                         "48 83 ec 40"
                                         "49 8b d8"
                                         "e8 X1 ? ? ? ? 48 8b f8 /d4"_xorpattern;
```
This brings me 3 bytes short of the target 
(`00007FFFBAC610F0 <ntdll.RtlpLookupPrimaryFunc | 48:83EC 28| sub     rsp,28  |`) - 3 bytes
Do i not understand something about "instruction length" or where to put `X`?
Please, help me figure this out! I know that using `x4` in the pattern works, but i want to understand!
many thanks 🙉 

If possible, also an example for cmp rel
![x64dbg_zZMgUiUpMN](https://github.com/user-attachments/assets/4ba7014e-cdcc-4819-bc84-4cc63748e702)

**Comment by scizzydo** [comment]
The X4 signifies the length of the remaining bytes to the end of the instruction. Since a call is 5 bytes, E8 and the int32 relative address, that's why there is 4. int32 being 4 bytes (address of X + 4 to get to the next instruction). In the instance of X1, you're saying the remaining instruction is 1 byte, so addr of X + 1. Like you stated of address of E8 + 5 + 52, its address of the next byte after E8 + 4, plus the dereferenced integer value (52).

TLDR: X goes at where you want to get the address start. The number following is the remaining length of the instruction at which X was.

Does this help?

**Comment by po0p** [comment]
> Does this help?

Hey, yeah, a little bit, still confused

>remaining bytes to the end of the instruction.

should it be called "operand length" then? And if so, it makes it even more confusing, since that operand size already goes to /d flag? Because
> The number following is the remaining length of the instruction at which X was.

Isnt this (well, talking windows x64 here) almost always 4 bytes? (talking about most often instructions decoded this way - with displacement relative to next instruction)

And, if possible, could you provide example pattern for rel cmp (REX.W + 39 /r) (https://www.felixcloutier.com/x86/cmp)
```
00007FF77986FDBE        | 48:83EC 40                      | sub     rsp,40                                             |
00007FF77986FDC2        | 48:8BF1                         | mov     rsi,rcx                                            |
00007FF77986FDC5        | 33ED                            | xor     ebp,ebp                                            |
00007FF77986FDC7        | 48:392D 7A032203                | cmp     qword ptr ds:[7FF77CA90148],rbp                    |
```

"48 8b f1 33 ed 48 39 2d x4 ? ? ? ?/d4"?

And here, which is same asm, but that is how most of rel-jmp-cmp patterns get generated by scanners (using bytes AFTER the jmp/call for identification
```
00007FF77986FDC7        | 48:392D 7A032203                | cmp     qword ptr ds:[7FF77CA90148],rbp                    |
00007FF77986FDCE        | 0F85 6A010000                   | jne     calc.7FF77986FF3E                           |
00007FF77986FDD4        | 41:BE 3C010000                  | mov     r14d,13C                                           |
00007FF77986FDDA        | 6548:8B0425 58000000            | mov     rax,qword ptr gs:[58]                              |
00007FF77986FDE3        | 48:8B10                         | mov     rdx,qword ptr ds:[rax]                             |
```

"48 39 2d x4 ? ? ? ? 0f 85 ? ? ? ? 41 be 3c /d4"?


also, how would i go when converting "old school" style patterns like these:
```
"sig" => "B9 B9 D8 D1 AE",
"off" => -0x149
```
(sig is in the middle of the function and is unique across binary, but function starts at -0x149 from &sig[0])

**Comment by scizzydo** [comment]
I see what you're asking. And the main answer is the reason those are there now with that, is it's adapted over time and old code just persisted. I used to have a length disassembler part of it (from nmd, look at first git history) and back then, it would get the length of each instruction up until where the offset was. In this scenario, address + instrlen + offset was calculated from the X offset, the length disassembler and the start of the pattern. The reason those are there now is because of just old code and adjustments over time without even thinking about it all. 

By default it also didn't have size to use (i.e. /d4), and also, by default it is set to 4, so /d4 isn't needed, just /d for dereferencing. Originally the 1/2/4/8 there was just for the relative address (which didn't exist in my code before pushing it to github). It wasn't intended for the /d flag and existed prior to separating out the length disassembler. But yes... I could rename things if it was so detrimental and caused issues...

> "sig" => "B9 B9 D8 D1 AE",
"off" => -0x149

In this case, I'd just subtract 0x149 from the address in my C++ code. I'm sure you can see, but there isn't anything for handling that in here without maybe making another flag for it. This code was just posted to share what I had, and something I don't actively go back into almost ever. There's plenty of things to adjust if need be. In my mind, I feel removing the size_ variable, keeping the 1/2/4/8 after the X offset for operand length, and then a flag like o:-0x149 to do your case mentioned above. 

Why haven't I done any of this? Well, because I haven't had a need to, and I don't personally have any of these situations ever come up warranting the need for it. In my code, if it's the intel instruction set, I just use /d (or /da for faster scans). I feel the offset subtraction of 0x149 isn't that reliable also, because what if during a future update, that same function still has those sig bytes, but an instruction between was modified, no longer making the start at -0x149?

**Comment by po0p** [comment]
Hey, thanks for longpost :)
This is not a feature request anyway, but thanks for explanation!
I will maybe make another flag, my usecase involves both constexpr patterns (say, finding libcurl functions) and some "dynamic" patterns coming from remote server. 
Or will continue looking if someone made a magical solution involving everything i need ;)

about `-0x149` -type patterns - it is actually produced by https://github.com/kweatherman/sigmakerex, and it is a great optimization for speed if you do a lot of pattern scanning (say, a game engine without any rtti at all, where pattern scanning bunch of stuff is the last resort).
Sometimes finding a function by pattern scanning for starting bytes can be very slow, if binary has a lot of similar functions, consider like 6 `push reg` at the start, your pattern is already 12 bytes, something like `48 8b 48 8c 48 8d 48 8a 48 aa 48 bb 48 cc 48 dd 48 ad 48 da` (mockup), and those also frequently change (not in length in my case) if underlying c++ classes change. 
considering you are doing plain old byte-by-byte scan (and not some avx magic (https://github.com/qis/signature)), time complexity directly depends on pattern length as far as i understand it.
Anyway, thanks for explanation ;)

**Comment by scizzydo** [comment]
I liked the ideas still. I just made a push for the stoi to handle - numbers. Didn't add the offset thing like you said, but I did strip out the # from the following flags. Defaults to 4 and can be set with X1,X2,X4,X8. The stoi you can make a new flag for and it should handle negative or positive, base 10 or base 16 numbers. I have never sought a plan to make it any more optimized either, as on my PC on the startup of application I scan 70 different patterns and it only takes .5 seconds

**Comment by po0p** [comment]
Hey, nice work, i will look into it tomorrow

**Comment by scizzydo** [comment]
I actually remember now too why I did have them different... just hit the use case in my app 😂 

```
.text:00000001403098D6 83 3D EF 92 19 04 00                    cmp     dword ptr cs:qword_1444A2BCC, 0
```
my pattern is like:
"83 3D X5 ? ? ? ? 0"
And since deref would be the operand size of 4, but the instruction from offset is 5, that was the use case of the two separated... about to revert the code

**Comment by po0p** [comment]
Oh, i see. I understand now. X<number> was neither `instruction size` neither `operand size`, it was `remaining instruction size`
1444A2BCC = 1403098D6 + 41992EF + 7
if we move x in place of operand start, the remaining instruction bytes are 5, b/c cmp takes 2 operands
so lib would calculate 1403098D6 + 41992EF + 2( x_pos =  2) +5 (operand of X) => correct

Btw, i also found this if you want to take a look. Really hard to understand what is going on (my poor head), but looks very sugar-ish. sadly no runtime parsing support.
https://github.com/moonshadow565/single_header/blob/master/include/ppp.hpp

About absolute offsets (-/+0x149) - i think i will just try to add 2 flags for positive/negative shift; or maybe just 1 if your compiletime stoi is good (havent tested).

**Comment by po0p** [comment]
Hey, i implemented the abs offset if you want to take a look (although i mashed everything into single header cuz thats easier for me to handle later)
Works pretty great!
https://github.com/scizzydo/PatternScanner/commit/485104fca256881a610696d161cf408436254ec4
in current project i have:
4 patterns with abs offset (4 curl functions basically) with _xorpattern)
at runtime:
1 rel call (E8 x4 ? ? ? ?)
2 abs offset (middle of func pattern -0x149)
1 rel cmp (
`48 39 2D 7A 03 22 03|  cmp     cs:qword_1432E0148, rbp ;`
) - still 4 remaining size thankfully, because its cmp with register, and it works!
thanks for help :)
